allennlp/models/reading_comprehension/dialog_qa.py

```diff
@@ -0,0 +1,407 @@
+        encoding_dim = encoded_question.size(-1)
+
+        # Shape: (batch_size * max_qa_count, passage_length, question_length)
+        passage_question_similarity = self._matrix_attention(repeated_encoded_passage, encoded_question)
+        # Shape: (batch_size * max_qa_count, passage_length, question_length)
+        passage_question_attention = util.last_dim_softmax(passage_question_similarity, question_mask)
+
+        # Shape: (batch_size * max_qa_count, passage_length, encoding_dim)
+        passage_question_vectors = util.weighted_sum(encoded_question, passage_question_attention)
+
+        # We replace masked values with something really negative here, so they don't affect the
+        # max below.
+        masked_similarity = util.replace_masked_values(passage_question_similarity,
+                                                       question_mask.unsqueeze(1),
+                                                       -1e7)
+
+        passage_mask = passage_mask.unsqueeze(1).repeat(1, max_qa_count, 1).view(batch_size * max_qa_count, -1)
+        question_passage_similarity = masked_similarity.max(dim=-1)[0].squeeze(-1)
+        question_passage_attention = util.last_dim_softmax(question_passage_similarity,
+                                                           passage_mask)
+        # Shape: (batch_size * max_qa_count, encoding_dim)
+        question_passage_vector = util.weighted_sum(repeated_encoded_passage, question_passage_attention)
+        tiled_question_passage_vector = question_passage_vector.unsqueeze(1).expand(batch_size * max_qa_count,
+                                                                                    passage_length,
+                                                                                    encoding_dim)
+
+        # Shape: (batch_size * max_qa_count, passage_length, encoding_dim * 4)
+        final_merged_passage = torch.cat([repeated_encoded_passage,
+                                          passage_question_vectors,
+                                          repeated_encoded_passage * passage_question_vectors,
+                                          repeated_encoded_passage * tiled_question_passage_vector],
+                                         dim=-1)
+
+        final_merged_passage = F.relu(self._merge_atten(final_merged_passage))
+
+        merged_passage_for_residual = self._variational_dropout(final_merged_passage)
+        residual_layer = self._variational_dropout(self._residual_encoder(merged_passage_for_residual, passage_mask))
+        self_atten_matrix = self._self_atten(residual_layer, residual_layer)
+
+        mask = passage_mask.resize(batch_size * max_qa_count,
+                                   passage_length,
+                                   1) * passage_mask.resize(batch_size * max_qa_count, 1, passage_length)
+
+        # torch.eye does not have a gpu implementation, so we are forced to use the cpu one and .cuda()
```

Comment:
`torch.eye` now has a GPU implementation.  So, to get the mask you want, you can just do this:

```python
passage_self_mask = torch.eye(passage_length, device=self_attention_matrix.device).unsqueeze(0).unsqueeze(0)
self_attention_mask = (1 - passage_self_mask) * passage_mask.unsqueeze(-1) * passage_mask.unsqueeze(-2)
```

Those two lines _should_ work to replace your current lines 215 through 225 (assuming you've undone the shape mangling, so you have `(batch_size, num_questions, ...)` shapes).